lib/phase/functional.js

```diff
@@ -0,0 +1,164 @@
+
+    return jobNames;
+}
+
+/**
+ * Make sure the Matrix they specified is valid
+ * @method validateJobMatrix
+ * @param  {Object}      job     Job to inspect
+ * @param  {String}      jobName Name of job (for error reporting)
+ * @return {Array}               List of errors
+ */
+function validateJobMatrix(job, jobName) {
+    let matrixSize = 1;
+    const errors = [];
+    const matrix = Hoek.reach(job, 'matrix', {
+        default: {}
+    });
+
+    Object.keys(matrix).forEach((row) => {
+        matrixSize *= matrix[row].length;
+    });
+
+    if (matrixSize > MAX_PERMUTATIONS) {
+        errors.push(`Job ${jobName}: Matrix cannot contain >${MAX_PERMUTATIONS} permutations`);
+    }
+
+    return errors;
+}
+
+/**
+ * Ensure the job is something that can be run
+ *  - has at least one step
+ *  - the image name
+ *  - not too many environment variables
+ * @method validateJobSchema
+ * @param  {Object}          doc Document that went through flattening
+ * @return {Array}               List of errors
+ */
+function validateJobSchema(doc) {
+    let errors = [];
+
+    // Jobs
+    const SCHEMA_JOB = Joi.object()
+        .keys({
+            steps: Joi.object().min(1).required(),
+            image: Joi.string().required(),
+            environment: Joi.object().max(MAX_ENVIRONMENT_VARS).default({})
+        })
+        .unknown(true)
+        // Add documentation
+        .options({
+            language: {
+                object: {
+                    min: 'requires at least one step',
+                    max: `can only have ${MAX_ENVIRONMENT_VARS} environment variables defined`
+                }
+            }
+        });
+
+    // Validate jobs contain required minimum fields
+    Object.keys(doc.jobs).forEach((jobName) => {
+        try {
+            Joi.attempt(doc.jobs[jobName], SCHEMA_JOB, `Job ${jobName}`);
+        } catch (err) {
+            errors.push(err.message);
+        }
+
+        errors = errors.concat(validateJobMatrix(doc.jobs[jobName], jobName));
+    });
+
+    return errors;
+}
+
+/**
+ * Get the defined workflow, or if missing generate one (list of all jobs in series)
+ * @method generateWorkflow
+ * @param  {Object}          doc Document that went through flattening
+ * @return {Array}               Workflow of jobs
+ */
+function generateWorkflow(doc) {
+    // Default to the list of jobs
+    return Hoek.reach(doc, 'workflow', {
+        default: getDefaultWorkflow(doc.jobs)
+    });
+}
+
+/**
+ * Ensure the workflow is a valid one
+ *  - Contains all defined jobs (no extra or missing)
+ *  - Main is first
+ * @method validateWorkflow
+ * @param  {Object}          doc Document that went through flattening
+ * @return {Array}               List of errors
+ */
+function validateWorkflow(doc) {
+    const errors = [];
+
+    // Validate that main is not included
+    if (doc.workflow.indexOf('main') > -1) {
+        errors.push('Workflow: "main" is implied as the first job and must be excluded');
```

Comment:
Should we just return errors here? 
Failing here means the next check is going to fail as well, and it will be confusing to users when they see both of these error message.